sstables/sstable_set.hh

```diff
@@ -47,24 +47,86 @@ public:
 // Structure holds all sstables (a.k.a. fragments) that belong to same run identifier, which is an UUID.
 // SStables in that same run will not overlap with one another.
 class sstable_run {
-    sstable_list _all;
+    std::unordered_set<shared_sstable> _all;
 public:
+    sstable_run(std::unordered_set<shared_sstable> all) : _all(std::move(all)) {}
     void insert(shared_sstable sst);
     void erase(shared_sstable sst);
     // Data size of the whole run, meaning it's a sum of the data size of all its fragments.
     uint64_t data_size() const;
-    const sstable_list& all() const { return _all; }
+    const std::unordered_set<shared_sstable>& all() const { return _all; }
 };
 
-class sstable_set {
+class versioned_sstable_set_data {
     std::unique_ptr<sstable_set_impl> _impl;
-    schema_ptr _schema;
+    std::unordered_map<utils::UUID, std::unordered_set<shared_sstable>> _all_runs;
+    std::unordered_map<shared_sstable, unsigned> _times_added;
+    std::set<shared_sstable> _all;
```

Comment:
Can we merge _all and _times_added? Using std::map<shared_sstable, unsigned>. I'm more concerned about exception safety and complexity than performance. I did not review the implementation yet to see if there's a problem.

Reply:
We can, I was thinking that I might be better to separate these two for easier understanding. In particular, I couldn't come up with a name that conveys the meanings of both these containers, I might end up with something like "_all_sstables_and_times_they_were_added"

Reply:
I would also favour consolidating the different containers into one. `_all` is a good name, as long as the value's type name tells us what it is.

Reply:
Maybe we can even add `_removed` here and have a value be a:
```
struct sstable_meta {
     shared_sstable sst;
     unsigned times_added;
     bool removed;
};
```

Reply:
I've combined them to a single std::map